air/src/trace/rows.rs

```diff
@@ -0,0 +1,349 @@
+#[derive(Debug, Copy, Clone, Eq, Ord, PartialOrd)]
+pub struct RowIndex(u32);
+
+impl Display for RowIndex {
+    fn fmt(&self, f: &mut Formatter) -> core::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+// FROM ROW INDEX
+// ================================================================================================
+
+impl From<RowIndex> for u32 {
+    fn from(step: RowIndex) -> u32 {
+        step.0
+    }
+}
+
+impl From<RowIndex> for u64 {
+    fn from(step: RowIndex) -> u64 {
+        step.0 as u64
+    }
+}
+
+impl From<RowIndex> for usize {
+    fn from(step: RowIndex) -> usize {
+        step.0 as usize
+    }
+}
+
+impl From<RowIndex> for Felt {
+    fn from(step: RowIndex) -> Felt {
+        Felt::from(step.0)
+    }
+}
+
+// INTO ROW INDEX
+// ================================================================================================
+
+/// Converts a usize value into a [`RowIndex`].
+///
+/// # Panics
+///
+/// This function will panic if the number represented by the usize is greater than the maximum
+/// [`RowIndex`] value, `u32::MAX`.
+impl From<usize> for RowIndex {
+    fn from(value: usize) -> Self {
+        let value = u32::try_from(value).map_err(|_| RowIndexError::InvalidSize(value)).unwrap();
+        value.into()
+    }
+}
+
+/// Converts a u64 value into a [`RowIndex`].
+///
+/// # Panics
+///
+/// This function will panic if the number represented by the u64 is greater than the maximum
+/// [`RowIndex`] value, `u32::MAX`.
+impl TryFrom<u64> for RowIndex {
+    type Error = RowIndexError<u64>;
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        let value = u32::try_from(value).map_err(|_| RowIndexError::InvalidSize(value))?;
+        Ok(RowIndex::from(value))
+    }
+}
+
+impl From<u32> for RowIndex {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+
+/// Converts an i32 value into a [`RowIndex`].
+///
+/// # Panics
+///
+/// This function will panic if the number represented by the i32 is less than 0.
+impl From<i32> for RowIndex {
+    fn from(value: i32) -> Self {
+        let value = u32::try_from(value).map_err(|_| RowIndexError::InvalidSize(value)).unwrap();
+        RowIndex(value)
+    }
+}
+
+// ROW INDEX OPS
+// ================================================================================================
+
+/// Subtracts a usize from a [`RowIndex`].
+///
+/// # Panics
+///
+/// This function will panic if the number represented by the usize is greater than the maximum
+/// [`RowIndex`] value, `u32::MAX`.
+impl Sub<usize> for RowIndex {
+    type Output = RowIndex;
+
+    fn sub(self, rhs: usize) -> Self::Output {
+        let rhs = u32::try_from(rhs).map_err(|_| RowIndexError::InvalidSize(rhs)).unwrap();
+        RowIndex(self.0 - rhs)
+    }
+}
+
+impl Sub<RowIndex> for usize {
+    type Output = RowIndex;
+
+    fn sub(self, rhs: RowIndex) -> Self::Output {
+        (self - rhs.0 as usize).into()
+    }
+}
```

Comment:
I'm curious: are there many places where we subtract row index from something?

Reply:
Just here
```
let current_shift = NUM_DECOMP_BITS * (OP_CYCLE_LEN - cycle_row - 1);
```